### PR TITLE
Parallel getpage requests and walredo pool

### PR DESCRIPTION
**Pull request by hlinnaka**
1st commit is a rebase of https://github.com/neondatabase/neon/pull/1563, to have a pool of walredo processes. That should all but eliminate the walredo process as a bottleneck for parallelism.

The 2nd commit adds parallelism or pipelining to how prefetched requests are processed. On 'main', we read one request from the client, process it, and send response back. Then we repeat for next request. With this PR, we use tokio async to read requests from the client connection as they arrive, spawn a new tokio task for each request, and send the responses back as they finish (still maintaining the order). We now process up to 4 prefetched requests in parallel, from the same client connection.

In a nutshell, this should increase the throughput processed GetPage requests,  in a sequential scan where you have a deep queue of prefetched requests, . It doesn't help with latency of individual requests.

This isn't a great solution to our sequential scan performance issues, because we're still burning the same amount of CPU per request. We're just now able to do it in parallel. So a single postgres backend can keep up to 4 CPUs busy in the pageserver. So even if this gives us acceptable throughput and we proceed with this approach, we should continue to optimize, to reduce the CPU consumption.

### CODE REVIEW

**Review by knizhnik — COMMENTED** [review]
This PR contains two changes: parallel processing of get page requests and walredo pool,
Them seems to be unrelated and also we have separate PR for walredo pool (#1563).
I still think that maintain pool of walredo process per tenant is not good idea: I think that we should  have global pool of walredo processes which make it possible to limit number of walredo processes and still use more than one walredo process for tenant if there are no other activity.

But processing each getpage request in separate tokio task is really cool.
Some performance numbers:

| branch | time of Q1 (sec) |
| --------- | ---------------------- |
| main    | 30                       |
| #2880 (this PR) |  13 |
| walredo_channel | 17.5    |
| async walredo | 22 |
| walredo_channel + parallel getpage | 14.5 |
| #1563 (walredo pool) | 18 |

So, as you see, there are to ways of increasing walredo speed:

1. spawn more walredo instances
2. async processing of walredo requests (buffering)

Both provides almost the same performance but certainly it is better when we can achieve the same speed with one walredo processes rather than with 4.

There is one more alternative: try to reduce walredo latency by replacing pipe with shared memory (this is what I currently try to implement).

So I prefer to split this PR into two and merge part with parallel getpage requests.
And concerning pool of walredo processes, I think that we should have global one, especially taken in account that tokio implementation of walredo communication is noticeable slower than sync one.

**Comment by hlinnaka** [comment]
> I still think that maintain pool of walredo process per tenant is not good idea: I think that we should have global pool of walredo processes which make it possible to limit number of walredo processes and still use more than one walredo process for tenant if there are no other activity.

We cannot share walredo processes across tenants for security reasons. So it has to be a per-tenant pool. Possibly with some global accounting and logic to shrink the pools of tenants if the total number of processes across all tenants is too large.

**Comment by knizhnik** [comment]
Yes, definitely we can not shared processes.
But we can limit number of processes.
Assume that we have pool with 100 walredo processes.
If some new tenant need to perform walredo and do not have associated process, then we shutdown one of the running processes and spawn new one.
If only one tenant is active and perform a lot of concurrent queries (or OLAP quyer with parallel plan), then we can assign more than one walredo process to the tenant.

So if pageserver is serving 1000 tenants, there will be no 1000 or even 4000 processes, but just 100.

May be we can also find some fast way of spawning new process, so that it has not to perform startup action.
For example we can use fork() to copy memory. The question is what to do with local files... May be there is also some way to "snapshot" it or prevent it fro unsatisfied changes.

**Comment by knizhnik** [comment]
Please notice that initially all walredo processes are the same: they do not have any tenant specific.
So if we can provide fast cloning or resetting of waledo process environment, we do not need to keep running walredo process for each tenant.

**Comment by hlinnaka** [comment]
> Some performance numbers:

@knizhnik just to be sure we're on the same page:

- all the numbers in this table are with prefetching enabled?
- the baseline when all data is in shared buffer cache is 4 s ?

**Comment by knizhnik** [comment]
> > Some performance numbers:
> 
> @knizhnik just to be sure we're on the same page:
> 
>     * all the numbers in this table are with prefetching enabled?
> 
>     * the baseline when all data is in shared buffer cache is 4 s ?

Yes, it is results with prefetch_buffers=100 and max_parallel_workers_per_gather=8.
And baseline with local cache is 4 seconds.

**Comment by koivunej** [comment]
Implemented a "lockless" version of the pooling in the async-walredo branch: https://github.com/neondatabase/neon/pull/2875/commits/7d800a4da5784c27bbab394e4e9822744efc07d4, so that there we can get the maximum benefit out of the mpmc queue.

File: pageserver/src/page_service.rs [hunk 1]
```diff
@@ -298,66 +310,101 @@ impl PageServerHandler {
...
+                            // error message is enough
+                            error!("error reading relation or page version: {:?}", e);
+                            PagestreamBeMessage::Error(PagestreamErrorResponse {
+                                message: e.to_string(),
+                            })
+                        });
+                        let response: Bytes = response.serialize();
+                        (response, request_type, start_time.elapsed().as_secs_f64())
+                    };
+                    inprogress_requests.push_back(tokio::spawn(task));
```
Comment (koivunej) [inline pageserver/src/page_service.rs:404]: This is actually interesting use of spawn and futures ordered. Had not ever considered it, usually use future ordered to execute the tasks as well, but this makes sense.

Since I was going to propose using a joinset before I understood the meaning of FuturesOrdered here, one thing to consider is that on drop the tasks are not cancelled.

**Comment by koivunej** [comment]
Rebased the last commit as https://github.com/neondatabase/neon/pull/2875/commits/e9b2288808f7014fd80b52217ecb87f9eba00fd2 though I am not 100% of how successful it was.
